coroutines-extra/src/test/scala/AsyncAwaitTests.scala

```diff
@@ -0,0 +1,137 @@
+package org.coroutines.extra
+
+
+
+import org.coroutines._
+import org.scalatest._
+import scala.concurrent.{ Await, Future }
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+
+
+
+class AsyncAwaitTests extends FunSuite with Matchers {
+  class TestException extends Throwable
+
+  /** Source: https://git.io/vorXv
+   *  The use of Async/Await as opposed to pure futures allows this control flow
+   *  to be written more easily.
+   *  The execution blocks when awaiting for the result of `f1`. `f2` only blocks
+   *  after `AsyncAwait.await(f1)` evaluates to `true`.
+   */
+  test("simple test") {
+    val future = AsyncAwait.async {
+      val f1 = Future(true)
+      val f2 = Future(42)
+      if (AsyncAwait.await(f1)) {
+        AsyncAwait.await(f2)
+      } else {
+        0
+      }
+    }
+    assert(Await.result(future, 1 seconds) == 42)
+  }
+
+  /** Asynchronous blocks of code can be defined either outside of or within any
+   *  part of an `async` block. This allows the user to avoid triggering the
+   *  computation of slow futures until it is necessary.
+   *  For instance, computation will not begin on `innerFuture` until
+   *  `await(trueFuture)` evaluates to true.
+   */
+  test("nested async blocks") {
+    val outerFuture = AsyncAwait.async {
+      val trueFuture = Future { true }
+      if (AsyncAwait.await(trueFuture)) {
+        val innerFuture = AsyncAwait.async {
+          AsyncAwait.await(Future { 100 } )
+        }
+        AsyncAwait.await(innerFuture)
+      } else {
+        200
+      }
+    }
+    assert(Await.result(outerFuture, 1 seconds) == 100)
+  }
+
+  /** Uncaught exceptions thrown inside async blocks cause the associated futures
+   *  to fail.
+   */
+  test("error handling test 1") {
+    val errorMessage = "System error!"
+    val exception = intercept[RuntimeException] {
+      val future = AsyncAwait.async {
+        sys.error(errorMessage)
+        AsyncAwait.await(Future("dog"))
+      }
+      val result = Await.result(future, 1 seconds)
+    }
+    assert(exception.getMessage == errorMessage)
+  }
+
+  test("error handling test 2") {
+    intercept[TestException] {
+      val future = AsyncAwait.async {
+        throw new TestException
+        yieldval((Future("god"), new AsyncAwait.Cell[String]))
```

Comment:
It's good that you brought this up. It's not clear to me if `yieldval`s should be allowed inside an `async` block. It is kind of an abstraction leak. Since you're already defining a new macro, perhaps you should use it to disallow `yieldval`.